internal/template-validator/tlsinfo/tlsinfo.go

```diff
@@ -174,31 +162,78 @@ func loadCertificates(directory string) (serverCrt *tls.Certificate, err error)
 	return &cert, nil
 }
 
-func (ti *TLSInfo) getCertificate() *tls.Certificate {
-	ti.certLock.Lock()
-	defer ti.certLock.Unlock()
-	return ti.cert
+func (ti *TLSInfo) updateTLSOptions(directory string) {
+	tlsOptions, err := loadTLSOptions(directory)
+
+	ti.lock.Lock()
+	defer ti.lock.Unlock()
+
+	if err != nil {
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = err
+		logger.Log.Error(err, "failed to load TLS options",
+			"directory", directory)
+		return
+	}
+
+	if tlsOptions.IsEmpty() {
+		// Using default configuration
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = nil
+		logger.Log.Info("TLS options are empty, using default configuration")
+		return
+	}
+
+	minVersion, err := tlsOptions.MinTLSVersionId()
+	if err != nil {
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = fmt.Errorf("TLS Configuration broken, min version misconfigured: %w", err)
+		logger.Log.Error(ti.tlsOptionsError, "TLS options are not valid",
+			"directory", directory)
+		return
+	}
+
+	ti.minTLSVersion = minVersion
+	ti.cipherSuites = common.CipherIDs(tlsOptions.OpenSSLCipherNames, nil)
+	ti.tlsOptionsError = nil
+	logger.Log.Info("TLS options retrieved", "directory", directory)
 }
 
-func (ti *TLSInfo) CreateTlsConfig() *tls.Config {
-	tlsConfig := &tls.Config{
-		GetCertificate: func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
-			cert := ti.getCertificate()
-			if cert == nil {
-				return nil, errors.New("no server certificate, server is not yet ready to receive traffic")
-			}
-			return cert, nil
-		},
+func loadTLSOptions(directory string) (*common.SSPTLSOptions, error) {
+	optionsFilePath := filepath.Join(directory, TLSOptionsFilename)
+	optionsJson, err := os.ReadFile(optionsFilePath)
+	if err != nil {
+		return nil, err
 	}
 
-	if !ti.sspTLSOptions.IsEmpty() {
-		tlsConfig.CipherSuites = common.CipherIDs(ti.sspTLSOptions.OpenSSLCipherNames, nil)
-		minVersion, err := ti.sspTLSOptions.MinTLSVersionId()
-		if err != nil {
-			panic(fmt.Sprintf("TLS Configuration broken, min version misconfigured %v", err))
-		}
-		tlsConfig.MinVersion = minVersion
+	tlsOptions := &common.SSPTLSOptions{}
+	if err = json.Unmarshal(optionsJson, tlsOptions); err != nil {
+		return nil, err
+	}
+
+	return tlsOptions, nil
+}
+
+func (ti *TLSInfo) CreateTlsConfig() (*tls.Config, error) {
+	ti.lock.Lock()
+	defer ti.lock.Unlock()
+
+	if ti.certError != nil {
+		return nil, fmt.Errorf("error getting certificate: %w", ti.certError)
+	}
+
+	if ti.tlsOptionsError != nil {
+		return nil, fmt.Errorf("error getting TLS options: %w", ti.tlsOptionsError)
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{*ti.cert},
```

Comment:
What if `ti.cert` is nil? You check `ti.certError` above but it might be worth to avoid a nil-ptr access by adding a second check.

Reply:
Same for `cipherSuites == nil`.

Reply:
In my opinion, it is better to panic in that case. If `ti.certError == nil` and `ti.cert == nil` , then it is a programmer's error and an error message is probably useless to the user. It's better that we catch the bug sooner if we see nil-ptr panic.

Reply:
Won't this result in a crash loop possibly? A Pod could otherwise appear as not Ready instead, which would look cleaner to me. Not going to block the PR though.

Reply:
You are right, I've added the check.